src/vivarium/framework/resource.py

```diff
@@ -0,0 +1,147 @@
+"""
+===================
+Resource Management
+===================
+
+This module provides a tool to manage dependencies on resources within a
+:mod:`vivarium` simulation. These resources take the form of things that can
+be created and utilized by components, for example columns in the
+:mod:`state table <vivarium.framework.population>`
+or :mod:`named value pipelines <vivarium.framework.values>`.
+
+Because these resources need to be created before they can be used, they are
+sensitive to ordering. The intent behind this tool is to provide an interface
+that allows other managers to register resources with the dependency manager
+and in turn ask for ordered sequences of these resources according to their
+dependencies or raise exceptions if this is not possible.
+
+"""
+from types import MethodType
+from typing import List, Any, Iterable
+import warnings
+
+import networkx as nx
+
+from vivarium.exceptions import VivariumError
+
+
+class ResourceError(VivariumError):
+    """Error raised when a dependency requirement is violated"""
+    pass
+
+
+RESOURCE_TYPES = {'value', 'value_source', 'value_modifier', 'column', 'stream'}
+
+
+class ResourceProducer:
+
+    def __init__(self, resource_type: str, resource_names: List[str], producer: MethodType, dependencies: List[str]):
+        self.resource_type = resource_type
+        self.resource_names = resource_names
+        self.producer = producer
+        self.dependencies = dependencies
+
+    def __repr__(self):
+        resources = ','.join([f'{self.resource_type}.{name}' for name in self.resource_names])
+        return f'ResourceProducer({resources})'
+
+
+class EmptySet:
+
+    def add(self, item: Any):
+        pass
+
+    def __contains__(self, item: Any) -> bool:
+        return False
+
+
+class ResourceGroup:
+
+    def __init__(self, phase: str, single_producer: bool = False):
+        self.phase = phase
+        # One initializer per component, maybe multiple value producers
+        self.producer_components = set() if single_producer else EmptySet()
+        self.resources = {}
+        self._graph = None
+
+    @property
+    def graph(self):
+        if self._graph is None:
+            self._graph = self._to_graph()
+        return self._graph
+
+    def add_resources(self, resource_type: str, resource_names: List[str],
+                      producer: MethodType, dependencies: List[str]):
+        if resource_type not in RESOURCE_TYPES:
+            raise ResourceError(f'Unknown resource type {resource_type}.  Permitted types are {RESOURCE_TYPES}.')
+        if resource_type == 'column':
```

Comment:
why are you only checking if resource_type == 'column'? Doesn't the EmptySet handle this so you can always do what's in this if block?

Reply:
This checks by resource type.  This is maybe a weird place to enforce this constraint, but the idea is that a single component can only provide one method that produces columns (population initializers) but may supply many methods that produce value modifiers/sources.

The same would be true if we were doing, e.g., time step handlers.  Each component should have one method that updates columns, but still has all the methods that source pipelines.